pkg/controller/uniteddeployment/uniteddeployment_controller.go

```diff
@@ -85,34 +104,167 @@ var _ reconcile.Reconciler = &ReconcileUnitedDeployment{}
+	}
+
+	nextReplicas, effectiveSpecifiedReplicas := GetAllocatedReplicas(nameToSubset, instance)
+	klog.V(4).Infof("Get UnitedDeployment %s/%s next replicas %v", instance.Namespace, instance.Name, nextReplicas)
+	if !effectiveSpecifiedReplicas {
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeSpecifySubbsetReplicas), "Specified subset replicas is ineffective")
+	}
+
+	nextPartitions := calcNextPartitions(instance, nextReplicas)
+	klog.V(4).Infof("Get UnitedDeployment %s/%s next partition %v", instance.Namespace, instance.Name, nextPartitions)
+
+	if err := r.manageSubsets(instance, nameToSubset, nextReplicas, nextPartitions, currentRevision, updatedRevision, subsetType); err != nil {
+		klog.Errorf("Fail to update UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeSubsetsUpdate), err.Error())
+		return reconcile.Result{}, nil
+	}
+
 	return reconcile.Result{}, nil
 }
+
+func (r *ReconcileUnitedDeployment) getNameToSubset(instance *appsv1alpha1.UnitedDeployment, control ControlInterface) (nameToSubset map[string]*Subset, err error) {
+	subSets, err := control.GetAllSubsets(instance)
+	if err != nil {
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeFindSubsets), err.Error())
+		return nil, fmt.Errorf("fail to get all Subsets for UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+	}
+
+	klog.V(4).Infof("Classify UnitedDeployment %s/%s by subSet name", instance.Namespace, instance.Name)
+	nameToSubsets := r.classifySubsetBySubsetName(instance, subSets)
+
+	nameToSubset, err = r.deleteDupSubset(instance, nameToSubsets, control)
+	if err != nil {
+		r.recorder.Event(instance.DeepCopy(), corev1.EventTypeWarning, fmt.Sprintf("Failed%s", eventTypeDupSubsetsDelete), err.Error())
+		return nil, fmt.Errorf("fail to manage duplicate Subset of UnitedDeployment %s/%s: %s", instance.Namespace, instance.Name, err)
+	}
+
+	return
+}
+
+func calcNextPartitions(ud *appsv1alpha1.UnitedDeployment, nextReplicas map[string]int32) map[string]int32 {
+	partitions := map[string]int32{}
+	for _, subset := range ud.Spec.Topology.Subsets {
+		var subsetPartition int32
+		if isManualUpdateStrategy(ud) && ud.Spec.Strategy.ManualUpdate != nil && ud.Spec.Strategy.ManualUpdate.Partitions != nil {
+			if partition, exist := ud.Spec.Strategy.ManualUpdate.Partitions[subset.Name]; exist {
+				subsetPartition = partition
+			}
+		}
+
+		if subsetReplicas, exist := nextReplicas[subset.Name]; exist && subsetPartition > subsetReplicas {
+			subsetPartition = subsetReplicas
+		}
+
+		partitions[subset.Name] = subsetPartition
+	}
+
+	return partitions
+}
+
+func isManualUpdateStrategy(ud *appsv1alpha1.UnitedDeployment) bool {
+	return len(ud.Spec.Strategy.Type) == 0 || ud.Spec.Strategy.Type == appsv1alpha1.ManualUpdateStrategyType
```

Comment:
The mutant webhook and default method should ensue the Type cannot be empty string.

Reply:
done